corehq/motech/repeaters/tasks.py

```diff
@@ -215,6 +226,210 @@ def _process_repeat_record(repeat_record):
+    """
+    Cycles through repeaters (repeatedly ;) ) until there are no more
+    repeat records ready to be sent.
+    """
+    while True:
+        yielded = False
+        for domain, repeater_id in iter_ready_repeater_ids_once():
+            if not domain_can_forward_now(domain):
+                continue
+            if rate_limit_repeater(domain, repeater_id):
+                continue
+
+            lock = get_repeater_lock(repeater_id)
+            lock_token = uuid.uuid1().hex  # The same way Lock does it
+            if lock.acquire(blocking=False, token=lock_token):
+                yielded = True
+                yield domain, repeater_id, lock_token
+
+        if not yielded:
+            # No repeaters are ready, or their domains can't forward or
+            # are paused.
+            return
+
+
+def iter_ready_repeater_ids_once():
+    """
+    Yields domain-repeater_id tuples in a round-robin fashion.
+
+    e.g. ::
+        ('domain1', 'repeater_id1'),
+        ('domain2', 'repeater_id2'),
+        ('domain3', 'repeater_id3'),
+        ('domain1', 'repeater_id4'),
+        ('domain2', 'repeater_id5'),
+        ...
+
+    """
+
+    def iter_domain_repeaters(dom):
+        try:
+            rep_id = repeater_ids_by_domain[dom].pop(0)
+        except IndexError:
+            return
+        else:
+            yield rep_id
+
+    repeater_ids_by_domain = get_repeater_ids_by_domain()
+    while True:
+        if not repeater_ids_by_domain:
+            return
+        for domain in list(repeater_ids_by_domain.keys()):
+            try:
+                repeater_id = next(iter_domain_repeaters(domain))
+            except StopIteration:
+                # We've exhausted the repeaters for this domain
+                del repeater_ids_by_domain[domain]
+                continue
+            else:
+                yield domain, repeater_id
+
+
+def get_repeater_lock(repeater_id):
+    redis = get_redis_connection()
+    name = f'process_repeater_{repeater_id}'
+    three_hours = 3 * 60 * 60
+    lock = Lock(redis, name, timeout=three_hours)
+    return MeteredLock(lock, name)
+
+
+def get_repeater_ids_by_domain():
+    repeater_ids_by_domain = Repeater.objects.get_all_ready_ids_by_domain()
+    always_enabled_domains = set(toggles.PROCESS_REPEATERS.get_enabled_domains())
+    return {
+        domain: repeater_ids
+        for domain, repeater_ids in repeater_ids_by_domain.items()
+        if (
+            domain in always_enabled_domains
+            # FeatureRelease toggle: Check whether domain is randomly enabled
+            or toggles.PROCESS_REPEATERS.enabled(domain, toggles.NAMESPACE_DOMAIN)
+        )
+    }
+
+
+@task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
+def process_repeater(domain, repeater_id, lock_token):
+
+    def get_task_signature(repeat_record):
+        task_ = {
+            State.Pending: process_pending_repeat_record,
+            State.Fail: process_failed_repeat_record,
```

Comment:
Just curious why move away from the `process_repeat_record` vs `retry_process_repeat_record`? I like that it is easy to understand the difference whereas these methods have the state context baked into the name, so someone has to understand how state works on repeat records.

Reply:
Ah these are happening in parallel of course, so we can't smoothly transition to continuing to use the existing method names.